### PR TITLE
feat: EXPOSED-60 Support json/json(b) column types

### PR DESCRIPTION
**Pull request by bog-walk**
Add 2 new column types: `JsonColumnType` & `JsonBColumnType` that can be used 3 ways:
```
json("column_name", Json, Type.serializer()) // or custom serializer
json<Type>("column_name", Json)
json("column_name", customEncoder(), customDecoder())
```

Dialects store JSON differently, usually in text or binary format, but may not reflect this in the name of the data type. Some may not even support the JSON data type, but use it as an alias for another type (e.g. [MariaDB](https://mariadb.com/kb/en/json-data-type/) maps JSON to LONGTEXT and [H2](https://www.h2database.com/html/datatypes.html#json_type) maps to `byte[]`). To allow for this variability, the user's choice between `json()` and `jsonb()` can be flexible depending on dialect, e.g.:

- [PostgreSQL] `json()` -> JSON, `jsonb()` -> JSONB
- [MySQL] [Only supports binary format](https://dev.mysql.com/doc/refman/8.0/en/json.html) but default type is still JSON, so both column types map to JSON (i.e. using `json()` and `jsonb()` is equivalent).
- [SQLite] No JSON type but allows JSON functions on TEXT columns, so `json()` maps to TEXT, while `jsonb()` throws.

**Current caveats:**
- Oracle and H2 are not yet supported.
- Use in DAO not fully supported.
- Only 2 functions currently supported: using paths/keys to extract from stored JSON as a scalar value/text or as a JSON object.

**Next steps:**
1. Extend support to Oracle and H2 (will likely mean overrides in `JsonBColumnType`).
2. Add a few more functions (assess most-used/important/broad functions across dialects).
3. Test and extend fully to DAO, as well as potential edge cases: e.g. raw Json strings, Json arrays or nested arrays, if field identifiers need escape formatting, indexing +/- misuse as primary/foreign key?
4. Assess potential use case for extending current cast types, literals, and parameters to include Json type.

### CODE REVIEW

File: exposed-core/build.gradle.kts [hunk 1]
```diff
@@ -12,5 +13,7 @@ dependencies {
     api(kotlin("stdlib"))
     api(kotlin("reflect"))
     api("org.jetbrains.kotlinx", "kotlinx-coroutines-core", Versions.kotlinCoroutines)
+    api("org.jetbrains.kotlinx", "kotlinx-serialization-json", Versions.kotlinxSerialization)
+    api("org.postgresql", "postgresql", Versions.postgre)
```
Comment (bog-walk) [inline exposed-core/build.gradle.kts:17]: PostgreSQL DB returns [a special type](https://jdbc.postgresql.org/documentation/publicapi/org/postgresql/util/PGobject.html) `PGobject` for deserialization. I didn't figure out a way to catch the type that doesn't involve including this specific dependency.

File: settings.gradle.kts [hunk 2]
```diff
@@ -15,5 +15,6 @@ include("exposed-crypt")
 pluginManagement {
     plugins {
         id("org.jetbrains.kotlin.jvm") version "1.8.22"
+        id("org.jetbrains.kotlin.plugin.serialization") version "1.8.22"
```
Comment (bog-walk) [inline settings.gradle.kts:18]: I believe that the serialization plugin must always match the kotlin plugin version. If true, does something need to be done with dependabot to ensure they don't diverge?

Comment (e5l) [reply settings.gradle.kts:18]: That's true

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/ColumnType.kt [hunk 3]
```diff
@@ -908,6 +909,53 @@ class EnumerationNameColumnType<T : Enum<T>>(
...
+        }
+        super.setParameter(stmt, index, parameterValue)
+    }
+}
+
+/**
+ * Column for storing JSON data in binary format.
+ */
+class JsonBColumnType<T : Any>(
+    /** Returns the function that encodes an object of type [T] to a JSON String. */
```
Comment (e5l) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/ColumnType.kt:951]: We can skip KDoc for parameters or write it as `@param` in the constructor or class KDoc

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt [hunk 4]
```diff
@@ -96,6 +98,32 @@ fun <T : Any?> ExpressionWithColumnType<T>.varPop(scale: Int = 2): VarPop<T> = V
...
+inline fun <reified T : Any> ExpressionWithColumnType<*>.jsonExtract(vararg path: String, toScalar: Boolean = true): JsonExtract<T> {
+    val columnType = when (T::class) {
+        String::class -> TextColumnType()
+        Boolean::class -> BooleanColumnType()
+        Long::class -> LongColumnType()
+        Int::class -> IntegerColumnType()
+        Short::class -> ShortColumnType()
+        Byte::class -> ByteColumnType()
+        Double::class -> DoubleColumnType()
+        Float::class -> FloatColumnType()
```
Comment (e5l) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt:119]: Should we add ByteArray as well?

**Review by e5l — APPROVED** [review]
Well done! LGTM

I would also add tests with a non-serializable class to check if it fails with a good message

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt (hunk 4 above)
Comment (bog-walk) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt:119]: What should ByteArray be mapped to? `BlobColumnType`? Or maybe `BasicBinaryColumnType`, since the length is unknown?

**Comment by xtexChooser** [comment]
thanks!
